LinkageCheckerTest.testFindLinkageProblems_shouldSuppressJvmCIPackage fails in Java 11

```
$ mvn test -Dtest=LinkageCheckerTest#testFindLinkageProblems_shouldSuppressJvmCIPackage 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 11.466 s <<< FAILURE! - in com.google.cloud.tools.opensource.classpath.LinkageCheckerTest
[ERROR] testFindLinkageProblems_shouldSuppressJvmCIPackage(com.google.cloud.tools.opensource.classpath.LinkageCheckerTest)  Time elapsed: 11.384 s  <<< FAILURE!
com.google.common.truth.AssertionErrorWithFacts: 
Missing classes from jdk.vm.ci should not be reported
value of: findLinkageProblems()
expected to be empty
but was : [
Class sun.text.normalizer.UCharacterProperty has default access and is referenced by com.oracle.svm.core.jdk8.Util_sun_text_normalizer_UCharacterProperty_JDK8 (different package),
Class sun.misc.JavaUtilZipFileAccess is not found referenced by ClassFile{location=com.oracle.substratevm:svm:19.2.0.1, className=com.oracle.svm.core.jdk8.zipfile.ZipFileUtil},
Class sun.misc.SharedSecrets is not found referenced by ClassFile{location=com.oracle.substratevm:svm:19.2.0.1, className=com.oracle.svm.core.jdk8.zipfile.ZipFileUtil},
Class sun.misc.PerfCounter is not found referenced by ClassFile{location=com.oracle.substratevm:svm:19.2.0.1, className=com.oracle.svm.core.jdk8.zipfile.ZipFileUtil},
Class sun.util.logging.LoggingSupport is not found referenced by ClassFile{location=com.oracle.substratevm:svm:19.2.0.1, className=com.oracle.svm.core.jdk.FormatAccessors},
sun.text.normalizer.UBiDiProps's method getSingleton() is not found referenced by ClassFile{location=com.oracle.substratevm:svm:19.2.0.1, className=com.oracle.svm.core.jdk8.Util_sun_text_normalizer_UBiDiProps_JDK8}]
	at com.google.cloud.tools.opensource.classpath.LinkageCheckerTest.testFindLinkageProblems_shouldSuppressJvmCIPackage(LinkageCheckerTest.java:922)

```
(I added new line characters for formatting)


Class | Comment
------------ | -------------
sun.text.normalizer.UCharacterProperty | Package sun.text.normalizer is in rt.jar in Java 8. Java 8 and Java 11 users cannot use it; only available to JVM internal code.
sun.misc.JavaUtilZipFileAccess | The class is removed in Java 11. Java 8 compiler warns about it.
sun.misc.SharedSecrets | The class is removed in Java 11. Java 8 compiler warns about it.
sun.misc.PerfCounter | The class is removed in Java 11. Java 8 compiler warns about it.
sun.util.logging.LoggingSupport | The class is in rt.jar in Java 8. Java 8 and Java 11 users cannot use it; only available to JVM internal code.
sun.text.normalizer.UBiDiProps | Package sun.text.normalizer is in rt.jar in Java 8. Java 8 and Java 11 users cannot use it; only available to JVM internal code.


# NullPointerException (Solved by #1826 )

```
suztomo-macbookpro44% which java
/Library/Java/JavaVirtualMachines/jdk-11-latest/Contents/Home/bin/java
suztomo-macbookpro44% mvn test -Dtest=LinkageCheckerTest#testFindLinkageProblems_shouldSuppressJvmCIPackage   
...
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 12.524 s <<< FAILURE! - in com.google.cloud.tools.opensource.classpath.LinkageCheckerTest
[ERROR] testFindLinkageProblems_shouldSuppressJvmCIPackage(com.google.cloud.tools.opensource.classpath.LinkageCheckerTest)  Time elapsed: 12.416 s  <<< ERROR!
java.lang.NullPointerException
	at com.google.cloud.tools.opensource.classpath.LinkageCheckerTest.testFindLinkageProblems_shouldSuppressJvmCIPackage(LinkageCheckerTest.java:922)
```

Stacktrace as per IntelliJ:

```
java.lang.NullPointerException
	at com.google.cloud.tools.opensource.classpath.InaccessibleClassProblem.toString(InaccessibleClassProblem.java:40)
	at java.base/java.lang.String.valueOf(String.java:2951)
	at java.base/java.lang.StringBuilder.append(StringBuilder.java:168)
	at java.base/java.util.AbstractCollection.toString(AbstractCollection.java:473)
	at com.google.common.truth.IterableSubject.actualCustomStringRepresentation(IterableSubject.java:110)
	at com.google.common.truth.Subject.butWas(Subject.java:1133)
	at com.google.common.truth.Subject.failWithActual(Subject.java:736)
	at com.google.common.truth.IterableSubject.isEmpty(IterableSubject.java:144)
	at com.google.cloud.tools.opensource.classpath.LinkageCheckerTest.testFindLinkageProblems_shouldSuppressJvmCIPackage(LinkageCheckerTest.java:922)
```

`JavaClass targetClass = classDumper.loadJavaClass(targetClassName);` works fine but `classDumper.findClassLocation("sun.text.normalizer.UCharacterProperty")` returning null. Where is the class then?

<img width="984" alt="Screen Shot 2020-11-19 at 15 38 52" src="https://user-images.githubusercontent.com/28604/99721511-5af1b280-2a7d-11eb-813d-643cafcf4041.png">


JDK 11 provides the class `sun.text.normalizer.UCharacterProperty`. It has default modifier:

<img width="1201" alt="Screen Shot 2020-11-19 at 15 45 23" src="https://user-images.githubusercontent.com/28604/99722164-3f3adc00-2a7e-11eb-8d0d-9e991b2dfb3f.png">




Java 8 provides the class too but with public modifier:

<img width="1203" alt="Screen Shot 2020-11-19 at 16 02 51" src="https://user-images.githubusercontent.com/28604/99723859-c6894f00-2a80-11eb-95fa-b9c837b9fdf9.png">

That's why the test has difference in Java 8 and Java 11; it's actually nothing to do with jvm ci package.



# sun.text.normalizer.UCharacterProperty

The class is not accessible even in Java 8 javac.

```
suztomo-macbookpro44% cat suztomo/Main.java
package suztomo;

import sun.text.normalizer.UCharacterProperty;

public class Main {
  public static void main(String[] arguments) {
    UCharacterProperty property = UCharacterProperty.getInstance();
    System.out.println("property.m_trieInitialValue_:" + property.m_trieInitialValue_);
  }
}

suztomo-macbookpro44% which javac  
/Library/Java/JavaVirtualMachines/jdk-8-latest/Contents/Home/bin/javac
suztomo-macbookpro44% javac suztomo/Main.java
suztomo/Main.java:3: error: package sun.text.normalizer does not exist
import sun.text.normalizer.UCharacterProperty;
                          ^
suztomo/Main.java:7: error: cannot find symbol
    UCharacterProperty property = UCharacterProperty.getInstance();
    ^
  symbol:   class UCharacterProperty
  location: class Main
suztomo/Main.java:7: error: cannot find symbol
    UCharacterProperty property = UCharacterProperty.getInstance();
                                  ^
  symbol:   variable UCharacterProperty
  location: class Main
3 errors
```

IntelliJ shows it's public in rt.jar.

<img width="952" alt="Screen Shot 2020-11-19 at 22 34 51" src="https://user-images.githubusercontent.com/28604/99754916-75488200-2ab7-11eb-99b6-949c035d36da.png">

https://discuss.gradle.org/t/package-sun-text-normalizer-does-not-exist/7354/9 says

> Seems like:
> Oracle’s ‘javac’ doesn’t grant access to some of the internal classes in ‘rt.jar’ which aren’t meant to be used by user code.

# sun.misc.SharedSecrets

The class is removed in Java 11 runtime.

```
suztomo-macbookpro44% cat suztomo/Main.java 
package suztomo;
import sun.misc.SharedSecrets;

public class Main {
  public static void main(String[] arguments) {
    SharedSecrets sharedSecrets = new SharedSecrets();
    System.out.println("SharedSecrets: " + sharedSecrets);
  }
}
```

Java 8 compiler warns about the future removal of the API.

```
suztomo-macbookpro44% java8
suztomo-macbookpro44% which java
/Library/Java/JavaVirtualMachines/jdk-8-latest/Contents/Home/bin/java
suztomo-macbookpro44% javac suztomo/Main.java
suztomo/Main.java:2: warning: SharedSecrets is internal proprietary API and may be removed in a future release
import sun.misc.SharedSecrets;
               ^
suztomo/Main.java:6: warning: SharedSecrets is internal proprietary API and may be removed in a future release
    SharedSecrets sharedSecrets = new SharedSecrets();
    ^
suztomo/Main.java:6: warning: SharedSecrets is internal proprietary API and may be removed in a future release
    SharedSecrets sharedSecrets = new SharedSecrets();
                                      ^
3 warnings
suztomo-macbookpro44% java suztomo.Main
SharedSecrets: sun.misc.SharedSecrets@7852e922
suztomo-macbookpro44% java11
suztomo-macbookpro44% which java
/Library/Java/JavaVirtualMachines/jdk-11-latest/Contents/Home/bin/java
suztomo-macbookpro44% java suztomo.Main
Exception in thread "main" java.lang.NoClassDefFoundError: sun/misc/SharedSecrets
	at suztomo.Main.main(Main.java:6)
Caused by: java.lang.ClassNotFoundException: sun.misc.SharedSecrets
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	... 1 more
```

# sun.misc.JavaUtilZipFileAccess

This is an interesting one where Java8-compiled class file does not work with Java 11.

```
suztomo-macbookpro44% cat suztomo/Main.java 
package suztomo;

import java.util.zip.ZipFile;
import sun.misc.JavaUtilZipFileAccess;

public class Main {
  public static void main(String[] arguments) {
    JavaUtilZipFileAccess access = new JavaUtilZipFileAccess() {
      @Override
      public boolean startsWithLocHeader(ZipFile zipFile) {
        return true;
      }
    };

    System.out.println("JavaUtilZipFileAccess available? " + access.startsWithLocHeader(null));
  }
}
```

Java 8 compiler warns about the future removal of this API.

```
suztomo-macbookpro44% javac suztomo/Main.java
suztomo/Main.java:3: warning: JavaUtilZipFileAccess is internal proprietary API and may be removed in a future release
import sun.misc.JavaUtilZipFileAccess;
               ^
suztomo/Main.java:7: warning: JavaUtilZipFileAccess is internal proprietary API and may be removed in a future release
    JavaUtilZipFileAccess access = null;
    ^
2 warnings
suztomo-macbookpro44% pwd
/Users/suztomo/hello-java8/src/main/java
suztomo-macbookpro44% javac suztomo/Main.java
suztomo/Main.java:4: warning: JavaUtilZipFileAccess is internal proprietary API and may be removed in a future release
import sun.misc.JavaUtilZipFileAccess;
               ^
suztomo/Main.java:8: warning: JavaUtilZipFileAccess is internal proprietary API and may be removed in a future release
    JavaUtilZipFileAccess access = new JavaUtilZipFileAccess() {
    ^
suztomo/Main.java:8: warning: JavaUtilZipFileAccess is internal proprietary API and may be removed in a future release
    JavaUtilZipFileAccess access = new JavaUtilZipFileAccess() {
                                       ^
3 warnings

```

It works in Java 8:

```
suztomo-macbookpro44% which java       
/Library/Java/JavaVirtualMachines/jdk-8-latest/Contents/Home/bin/java
suztomo-macbookpro44% java suztomo.Main
JavaUtilZipFileAccess available? true
```

It throws `NoClassDefFoundError` with Java 11.

```
suztomo-macbookpro44% java11
suztomo-macbookpro44% which java
/Library/Java/JavaVirtualMachines/jdk-11-latest/Contents/Home/bin/java
suztomo-macbookpro44% java suztomo.Main
Error: Unable to initialize main class suztomo.Main
Caused by: java.lang.NoClassDefFoundError: sun/misc/JavaUtilZipFileAccess
```

# sun.misc.PerfCounter

The class is removed in Java 11.

```
suztomo-macbookpro44% cat suztomo/Main.java
package suztomo;
import sun.misc.PerfCounter;

public class Main {
  public static void main(String[] arguments) {
    PerfCounter counter = PerfCounter.getFindClasses();
    System.out.println("SharedSecrets: " + counter);
  }
}
```

The Java 8 compiler warns about the future removal.

```
suztomo-macbookpro44% java8
suztomo-macbookpro44% javac suztomo/Main.java
suztomo/Main.java:2: warning: PerfCounter is internal proprietary API and may be removed in a future release
import sun.misc.PerfCounter;
               ^
suztomo/Main.java:6: warning: PerfCounter is internal proprietary API and may be removed in a future release
    PerfCounter counter = PerfCounter.getFindClasses();
    ^
suztomo/Main.java:6: warning: PerfCounter is internal proprietary API and may be removed in a future release
    PerfCounter counter = PerfCounter.getFindClasses();
                          ^
3 warnings
suztomo-macbookpro44% java suztomo.Main      
SharedSecrets: sun.classloader.findClasses = 1
suztomo-macbookpro44% java11
suztomo-macbookpro44% java suztomo.Main
Exception in thread "main" java.lang.NoClassDefFoundError: sun/misc/PerfCounter
	at suztomo.Main.main(Main.java:6)
Caused by: java.lang.ClassNotFoundException: sun.misc.PerfCounter
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	... 1 more
```

# sun.util.logging.LoggingSupport

The class is unavailable in both Java 8 and Java 11.

```
suztomo-macbookpro44% cat suztomo/Main.java  
package suztomo;
import sun.util.logging.LoggingSupport;

public class Main {
  public static void main(String[] arguments) {
    System.out.println("LoggingSupport available: " + LoggingSupport.isAvailable());
  }
}
```

```
suztomo-macbookpro44% java8            
suztomo-macbookpro44% javac suztomo/Main.java
suztomo/Main.java:2: error: package sun.util.logging does not exist
import sun.util.logging.LoggingSupport;
                       ^
suztomo/Main.java:6: error: cannot find symbol
    System.out.println("LoggingSupport available: " + LoggingSupport.isAvailable());
                                                      ^
  symbol:   variable LoggingSupport
  location: class Main
2 errors
suztomo-macbookpro44% java11
suztomo-macbookpro44% javac suztomo/Main.java
suztomo/Main.java:2: error: package sun.util.logging does not exist
import sun.util.logging.LoggingSupport;
                       ^
suztomo/Main.java:6: error: cannot find symbol
    System.out.println("LoggingSupport available: " + LoggingSupport.isAvailable());
                                                      ^
  symbol:   variable LoggingSupport
  location: class Main
2 errors

```

# sun.text.normalizer.UBiDiProps

Package `sun.text.normalizer` is unavailable in Java8 or Java11.

```
suztomo-macbookpro44% cat suztomo/Main.java
package suztomo;
import sun.text.normalizer.UBiDiProps;

public class Main {
  public static void main(String[] arguments) {
    System.out.println("LoggingSupport available: " + UBiDiProps.getDummy());
  }
}
```

```
suztomo-macbookpro44% which javac
/Library/Java/JavaVirtualMachines/jdk-8-latest/Contents/Home/bin/javac
suztomo-macbookpro44% javac suztomo/Main.java
suztomo/Main.java:2: error: package sun.text.normalizer does not exist
import sun.text.normalizer.UBiDiProps;
                          ^
suztomo/Main.java:6: error: cannot find symbol
    System.out.println("LoggingSupport available: " + UBiDiProps.getDummy());
                                                      ^
  symbol:   variable UBiDiProps
  location: class Main
2 errors
suztomo-macbookpro44% java11
suztomo-macbookpro44% which javac
/Library/Java/JavaVirtualMachines/jdk-11-latest/Contents/Home/bin/javac
suztomo-macbookpro44% javac suztomo/Main.java
suztomo/Main.java:2: error: package sun.text.normalizer is not visible
import sun.text.normalizer.UBiDiProps;
               ^
  (package sun.text.normalizer is declared in module java.base, which does not export it)
1 error
```

--- a/dependencies/src/main/resources/linkage-checker-exclusion-default.xml
+++ b/dependencies/src/main/resources/linkage-checker-exclusion-default.xml
@@ -43,6 +43,68 @@
   </LinkageError>
   <LinkageError>
     <Target>
+      <Package name="sun.text.normalizer" />
+    </Target>
+    <Source>
+      <Package name="com.oracle.svm.core.jdk8" />
+    </Source>
+    <Reason>
+      Substratevm is part of GraalVM and the virtual machine references classes in the JVM-internal
+      package 'sun.text.normalizer' that are not available in JVM runtime.
+      https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1825
+    </Reason>
+  </LinkageError>
+  <LinkageError>
+    <Target>
+      <Class name="sun.util.logging.LoggingSupport" />
+    </Target>
+    <Source>
+      <Class name="com.oracle.svm.core.jdk.FormatAccessors" />
+    </Source>
+    <Reason>
+      Substratevm references the JVM-internal class 'sun.util.logging.LoggingSupport' that are not
+      available in JVM runtime.
+      https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1825
+    </Reason>
+  </LinkageError>
+  <LinkageError>
+    <Target>
+      <Class name="sun.misc.JavaUtilZipFileAccess" />
+    </Target>
+    <Source>
+      <Class name="com.oracle.svm.core.jdk8.zipfile.ZipFileUtil" />
+    </Source>
+    <Reason>
+      Substratevm references the class that are only available in Java 8.
+      https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1825
+    </Reason>
+  </LinkageError>
+  <LinkageError>
+    <Target>
+      <Class name="sun.misc.SharedSecrets" />
+    </Target>
+    <Source>
+      <Package name="com.oracle.svm.core.jdk8" />
+    </Source>
+    <Reason>
+      Substratevm references the class that are only available in Java 8.
+      https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1825
+    </Reason>
+  </LinkageError>
+  <LinkageError>
+    <Target>
+      <Class name="sun.misc.PerfCounter" />
+    </Target>
+    <Source>
+      <Package name="com.oracle.svm.core.jdk8" />
+    </Source>
+    <Reason>
+      Substratevm references the class that are only available in Java 8.
+      https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1825
+    </Reason>
+  </LinkageError>
+  <LinkageError>
+    <Target>
       <Class name="org.mockito.internal.creation.bytebuddy.MockMethodDispatcher" />
     </Target>
     <Source>

--- a/dependencies/src/main/resources/linkage-checker-exclusion-default.xml
+++ b/dependencies/src/main/resources/linkage-checker-exclusion-default.xml
@@ -49,7 +49,7 @@
       <Package name="com.oracle.svm.core.jdk8" />
     </Source>
     <Reason>
-      Substratevm is part of GraalVM and the virtual machine references classes in the JVM-internal
+      Substrate VM is part of GraalVM and the virtual machine references classes in the JVM-internal
       package 'sun.text.normalizer' that are not available in JVM runtime.
       https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1825
     </Reason>
@@ -62,7 +62,7 @@
       <Class name="com.oracle.svm.core.jdk.FormatAccessors" />
     </Source>
     <Reason>
-      Substratevm references the JVM-internal class 'sun.util.logging.LoggingSupport' that are not
+      Substrate VM references the JVM-internal class 'sun.util.logging.LoggingSupport' that are not
       available in JVM runtime.
       https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1825
     </Reason>
@@ -75,7 +75,7 @@
       <Class name="com.oracle.svm.core.jdk8.zipfile.ZipFileUtil" />
     </Source>
     <Reason>
-      Substratevm references the class that are only available in Java 8.
+      Substrate VM references the class that are only available in Java 8.
       https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1825
     </Reason>
   </LinkageError>
@@ -87,7 +87,7 @@
       <Package name="com.oracle.svm.core.jdk8" />
     </Source>
     <Reason>
-      Substratevm references the class that are only available in Java 8.
+      Substrate VM references the class that are only available in Java 8.
       https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1825
     </Reason>
   </LinkageError>
@@ -99,7 +99,7 @@
       <Package name="com.oracle.svm.core.jdk8" />
     </Source>
     <Reason>
-      Substratevm references the class that are only available in Java 8.
+      Substrate VM references the class that are only available in Java 8.
       https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1825
     </Reason>
   </LinkageError>

